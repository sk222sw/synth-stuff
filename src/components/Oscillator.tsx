import * as React from 'react'
import Knob from 'react-canvas-knob'
import styled from 'styled-components'
import { Row } from './styles/index'
import Waveforms from './Waveforms'

const StyledInputContainer = styled.div`
  display: flex;
  flex-direction: column;
  max-width: 90px;
  align-items: center;
`

const StyledInput = styled.input`
  text-align: center;
` as any

const Label = styled.span`
  font-size: 10px;
`

const StyledOscillator = styled.div`
  margin: 12px;
  padding: 12px;
  border: solid 1px rgba(0, 0, 0, 0.12);
  width: 250px;
`

interface Props {
  setOscillatorVolume: (value: number) => void
  setOffset: (oscillator: Oscillator) => (value: number) => void
  setSemi: (value: number) => void
  setWaveform: (value: string) => void
  oscillator: {}
  waveforms: string[]
}

const Oscillator = ({
  setOscillatorVolume,
  setOffset,
  setSemi,
  setWaveform,
  oscillator,
  waveforms,
}: Props) => (
  <StyledOscillator>
    <Row>
      <StyledInputContainer>
        <Knob
          value={Number(oscillator.peakVolume.toFixed(2))}
          max={1}
          min={0}
          step={0.001}
          onChange={setOscillatorVolume(oscillator)}
          onChangeEnd={setOscillatorVolume(oscillator)}
          width={28}
          height={28}
          thickness={0.3}
          angleOffset={-125}
          angleArc={250}
        />
        <Label>Volume</Label>
      </StyledInputContainer>
      <Waveforms
        waveforms={waveforms}
        onWaveformChange={setWaveform}
      />
      <StyledInputContainer>
        <StyledInput
          type="number"
          max="50"
          min="-50"
          value={oscillator.offset}
          onChange={(v: HtmlInEve) => setOffset(oscillator)(v.target.value)}
        />
        <Label>
          Cent
        </Label>
      </StyledInputContainer>
      <StyledInputContainer>
        <StyledInput
          type="number"
          max="24"
          min="-24"
          value={oscillator.semi}
          onChange={(event: any) => setSemi(oscillator)(event.target.value)}
        />
        <Label>
          Semi
        </Label>
      </StyledInputContainer>
    </Row>
</StyledOscillator>)

export default Oscillator
