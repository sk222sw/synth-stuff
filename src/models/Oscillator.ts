const start = (audioContext, oscillator, currentTime, envelope) => {
  if (!oscillator.playing) {
    oscillator.oscillator.connect(oscillator.gain)
    oscillator.gain.connect(audioContext.destination)

    const curr = audioContext.currentTime
    const { oscillator: oscillatorNode } = oscillator

    const { a: attack, d: decay, s: sustain } = envelope

    const attackEnd = curr + _millisecondToThousandOfASecond(attack)
    const decayEnd = attackEnd + _millisecondToThousandOfASecond(decay)

    oscillatorNode.start()
    oscillator.playing = true

    if (attack && !decay && !sustain) {
      scheduleAttack(oscillator, attackEnd)
    }
    if (!attack && decay && sustain) {
      scheduleDecayAndSustain(oscillator, sustain, attackEnd, decayEnd)
    }
    if (!attack && decay && !sustain) {
      scheduleDecay(oscillator, attackEnd, decayEnd)
    }
    if (attack && decay && !sustain) {
      scheduleAttack(oscillator, attackEnd)
      scheduleDecay(oscillator, attackEnd, decayEnd)
    }
    if (attack && decay && sustain) {
      scheduleAttack(oscillator, attackEnd)
      scheduleDecayAndSustain(oscillator, sustain, attackEnd, decayEnd)
    }
    if (!attack && !decay && sustain) {
      scheduleSustain(oscillator, sustain, curr)
    }
    if (attack && !decay && sustain) {
      scheduleAttack(oscillator, attackEnd)
      scheduleSustain(oscillator, sustain, attackEnd)
    }

  }

  return oscillator
}

const scheduleAttack = (oscillator, attackEnd) => {
  oscillator.gain.gain.value = 0
  _rampGainAndBeQuite(oscillator.gain, oscillator.volume, attackEnd)
}

const scheduleDecayAndSustain = (oscillator, sustain, attackEnd, decayEnd) => {
  oscillator.gain.gain.setTargetAtTime(oscillator.volume, attackEnd, 0)
  _rampGain(oscillator.gain, oscillator.volume * sustain, decayEnd)
}

const scheduleSustain = (oscillator, sustain, curr) => {
  oscillator.gain.gain.setTargetAtTime(oscillator.volume * sustain, curr, 0)
}

const scheduleDecay = (oscillator, attackEnd, decayEnd) => {
  oscillator.gain.gain.setTargetAtTime(oscillator.volume, attackEnd, 0)
  _rampGainAndBeQuite(oscillator.gain, 0, decayEnd)
}

export const _millisecondToThousandOfASecond = x => x / 1000

const stop = (oscillator, release) => {
  oscillator.gain.gain.cancelScheduledValues(0)
  oscillator.playing = false
  oscillator.oscillator.stop(release)
  _rampGain(oscillator.gain, 0.0, release)
}

const create = audioContext =>
  ({ frequency = 0, volume = 0, id = 0, offset = 0, waveform = 'sine', keyPress = 'a', semi = 0 } = {}) => {
    const oscillator = audioContext.createOscillator()
    const gain = audioContext.createGain()

    oscillator.frequency.value = Number(frequency)
    oscillator.detune.value = Number(semi * 100)
    oscillator.detune.value += Number(offset)

    oscillator.type = waveform

    gain.gain.value = volume

    return {
      oscillator,
      gain,
      id,
      frequency,
      offset,
      semi,
      keyPress,
      playing: false,
      volume,
      waveform,
    }
  }

const setVolume = (oscillator, volume) => {
  oscillator.volume = 2
  oscillator.gain.gain.value = volume
  return oscillator
}

const setFrequency = (oscillator, frequency) => {
  oscillator.frequency = frequency
  oscillator.oscillator.frequency.value = Number(oscillator.frequency)
  oscillator.oscillator.detune.value = Number(oscillator.semi * 100)
  oscillator.oscillator.detune.value += Number(oscillator.offset)
  return oscillator
}

const setOffset = (oscillator, offset) => {
  oscillator.offset = offset
  setFrequency(oscillator, oscillator.frequency)
  return oscillator
}

const setSemi = (oscillator, semi) => {
  oscillator.semi = semi
  setFrequency(oscillator, oscillator.frequency)
  return oscillator
}

const setWaveform = (oscillator, waveform) => {
  oscillator.waveform = waveform
  oscillator.oscillator.type = waveform
  return oscillator
}

export const _rampGain = (gainNode, gain, time) => {
  gainNode.gain.linearRampToValueAtTime(gain, time)
}

export const _rampGainAndBeQuite = (gainNode, gain, time, quiteAt = 0) => {
  gainNode.gain.linearRampToValueAtTime(gain, time)
  gainNode.gain.setTargetAtTime(0, time, quiteAt)
}

export default {
  start,
  stop,
  create,
  setFrequency,
  setVolume,
  setOffset,
  setSemi,
  setWaveform,
}
