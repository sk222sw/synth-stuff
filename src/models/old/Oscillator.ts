import * as R from 'ramda'
import { Effect } from './Effect'

export interface OscillatorFactory {
  audioContext: AudioContext
  oscillatorNode: OscillatorNode
  gainNode: GainNode
  effects: BiquadFilterNode[],
  start(time?: number): OscillatorFactory
  stop(time?: number): OscillatorFactory
  addEffect(effect: Effect): OscillatorFactory
  setFrequency(frequency: number): void
  setVolume(volume: number): void
  setupConnections(nodes: AudioNode[]): void
}

export const createOscillator =
  (audioContext: AudioContext) => ({
    frequency = 1000,
    volume = 0.5,
    playing = false,
    oscillatorNode = audioContext.createOscillator(),
    gainNode = audioContext.createGain(),
    effects = [] as Effect[],
  }) => ({
    setFrequency(fr) {
      oscillatorNode.frequency.value = fr
    },
    setVolume(value) {
      gainNode.gain.value = value
    },
    setupConnections(nodes: AudioNode[]) {
      const connectedNodes = nodes.reduce((prev, curr) => {
        if (prev.length) {
          R.last(prev).connect(curr)
        }
        return [...prev, curr]
      }, [] as AudioNode[])

      return connectedNodes
    },
    start(time = 0) {
      gainNode.gain.value = volume
      oscillatorNode.frequency.value = frequency

      const toConnect = [oscillatorNode, gainNode] as any

      effects.forEach(x => toConnect.push(x.effectNode))

      const connectedNodes =
        this.setupConnections(toConnect)

      R.last(connectedNodes).connect(audioContext.destination)

      oscillatorNode.start(time)
      return this
    },
    stop(time = 0) {
      oscillatorNode.stop()
      return this
    },
  })
