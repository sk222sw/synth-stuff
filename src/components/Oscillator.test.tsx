import { shallow } from 'enzyme'
import * as React from 'react'
import Oscillator from './Oscillator'

test('Oscillator', () => {
  const wrapper = shallow(
    <Oscillator
      oscillator={{
        offset: 0,
        playing: false,
        semi: 0,
        waveform: 'sine',
        peakVolume: 0.5,
      }}
      waveforms={[]}
      setOscillatorVolume={() => () => null}
      setOffset={() => () => null}
      setSemi={() => () => null}
      setWaveform={() => () => null}
    />,
  )

  expect(wrapper).toMatchSnapshot()
})
