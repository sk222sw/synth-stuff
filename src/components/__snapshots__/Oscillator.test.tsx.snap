// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Oscillator 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <styled.div>
    <styled.div>
        <styled.div>
            <Knob
                angleArc={250}
                angleOffset={-125}
                bgColor="#EEE"
                clockwise={true}
                cursor={false}
                disableMouseWheel={false}
                disableTextInput={false}
                displayInput={true}
                fgColor="#EA2"
                font="Arial"
                fontWeight="bold"
                height={28}
                inputColor=""
                lineCap="butt"
                log={false}
                max={1}
                min={0}
                onChange={[Function]}
                onChangeEnd={[Function]}
                readOnly={false}
                step={0.001}
                stopper={true}
                thickness={0.3}
                value={0.5}
                width={28}
            />
            <styled.span>
                Volume
            </styled.span>
        </styled.div>
        <Waveforms
            onWaveformChange={[Function]}
            waveforms={Array []}
        />
        <styled.div>
            <styled.input
                max="50"
                min="-50"
                onChange={[Function]}
                type="number"
                value={0}
            />
            <styled.span>
                Cent
            </styled.span>
        </styled.div>
        <styled.div>
            <styled.input
                max="24"
                min="-24"
                onChange={[Function]}
                type="number"
                value={0}
            />
            <styled.span>
                Semi
            </styled.span>
        </styled.div>
    </styled.div>
</styled.div>,
  "nodes": Array [
    <styled.div>
      <styled.div>
            <styled.div>
                  <Knob
                        angleArc={250}
                        angleOffset={-125}
                        bgColor="#EEE"
                        clockwise={true}
                        cursor={false}
                        disableMouseWheel={false}
                        disableTextInput={false}
                        displayInput={true}
                        fgColor="#EA2"
                        font="Arial"
                        fontWeight="bold"
                        height={28}
                        inputColor=""
                        lineCap="butt"
                        log={false}
                        max={1}
                        min={0}
                        onChange={[Function]}
                        onChangeEnd={[Function]}
                        readOnly={false}
                        step={0.001}
                        stopper={true}
                        thickness={0.3}
                        value={0.5}
                        width={28}
                  />
                  <styled.span>
                        Volume
                  </styled.span>
            </styled.div>
            <Waveforms
                  onWaveformChange={[Function]}
                  waveforms={Array []}
            />
            <styled.div>
                  <styled.input
                        max="50"
                        min="-50"
                        onChange={[Function]}
                        type="number"
                        value={0}
                  />
                  <styled.span>
                        Cent
                  </styled.span>
            </styled.div>
            <styled.div>
                  <styled.input
                        max="24"
                        min="-24"
                        onChange={[Function]}
                        type="number"
                        value={0}
                  />
                  <styled.span>
                        Semi
                  </styled.span>
            </styled.div>
      </styled.div>
</styled.div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Oscillator
        oscillator={
                Object {
                        "offset": 0,
                        "peakVolume": 0.5,
                        "playing": false,
                        "semi": 0,
                        "waveform": "sine",
                      }
        }
        setOffset={[Function]}
        setOscillatorVolume={[Function]}
        setSemi={[Function]}
        setWaveform={[Function]}
        waveforms={Array []}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "oscillator": Object {
            "offset": 0,
            "peakVolume": 0.5,
            "playing": false,
            "semi": 0,
            "waveform": "sine",
          },
          "setOffset": [Function],
          "setOscillatorVolume": [Function],
          "setSemi": [Function],
          "setWaveform": [Function],
          "waveforms": Array [],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <styled.div>
          <styled.div>
                    <styled.div>
                              <Knob
                                        angleArc={250}
                                        angleOffset={-125}
                                        bgColor="#EEE"
                                        clockwise={true}
                                        cursor={false}
                                        disableMouseWheel={false}
                                        disableTextInput={false}
                                        displayInput={true}
                                        fgColor="#EA2"
                                        font="Arial"
                                        fontWeight="bold"
                                        height={28}
                                        inputColor=""
                                        lineCap="butt"
                                        log={false}
                                        max={1}
                                        min={0}
                                        onChange={[Function]}
                                        onChangeEnd={[Function]}
                                        readOnly={false}
                                        step={0.001}
                                        stopper={true}
                                        thickness={0.3}
                                        value={0.5}
                                        width={28}
                              />
                              <styled.span>
                                        Volume
                              </styled.span>
                    </styled.div>
                    <Waveforms
                              onWaveformChange={[Function]}
                              waveforms={Array []}
                    />
                    <styled.div>
                              <styled.input
                                        max="50"
                                        min="-50"
                                        onChange={[Function]}
                                        type="number"
                                        value={0}
                              />
                              <styled.span>
                                        Cent
                              </styled.span>
                    </styled.div>
                    <styled.div>
                              <styled.input
                                        max="24"
                                        min="-24"
                                        onChange={[Function]}
                                        type="number"
                                        value={0}
                              />
                              <styled.span>
                                        Semi
                              </styled.span>
                    </styled.div>
          </styled.div>
</styled.div>,
        "_debugID": 2,
        "_renderedOutput": <styled.div>
          <styled.div>
                    <styled.div>
                              <Knob
                                        angleArc={250}
                                        angleOffset={-125}
                                        bgColor="#EEE"
                                        clockwise={true}
                                        cursor={false}
                                        disableMouseWheel={false}
                                        disableTextInput={false}
                                        displayInput={true}
                                        fgColor="#EA2"
                                        font="Arial"
                                        fontWeight="bold"
                                        height={28}
                                        inputColor=""
                                        lineCap="butt"
                                        log={false}
                                        max={1}
                                        min={0}
                                        onChange={[Function]}
                                        onChangeEnd={[Function]}
                                        readOnly={false}
                                        step={0.001}
                                        stopper={true}
                                        thickness={0.3}
                                        value={0.5}
                                        width={28}
                              />
                              <styled.span>
                                        Volume
                              </styled.span>
                    </styled.div>
                    <Waveforms
                              onWaveformChange={[Function]}
                              waveforms={Array []}
                    />
                    <styled.div>
                              <styled.input
                                        max="50"
                                        min="-50"
                                        onChange={[Function]}
                                        type="number"
                                        value={0}
                              />
                              <styled.span>
                                        Cent
                              </styled.span>
                    </styled.div>
                    <styled.div>
                              <styled.input
                                        max="24"
                                        min="-24"
                                        onChange={[Function]}
                                        type="number"
                                        value={0}
                              />
                              <styled.span>
                                        Semi
                              </styled.span>
                    </styled.div>
          </styled.div>
</styled.div>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Oscillator
    oscillator={
        Object {
            "offset": 0,
            "peakVolume": 0.5,
            "playing": false,
            "semi": 0,
            "waveform": "sine",
          }
    }
    setOffset={[Function]}
    setOscillatorVolume={[Function]}
    setSemi={[Function]}
    setWaveform={[Function]}
    waveforms={Array []}
/>,
}
`;
